lib/server.js

```diff
@@ -84,11 +84,30 @@ function createServer (options, httpHandler) {
 
         multipleBindings.call(this, server, httpHandler, options, listenOptions, () => {
           this[kState].listening = true
+          this[kAvvioBoot](() => {})
+          hookRunnerApplication('onListen', this[kAvvioBoot], this, function (err) {
+            if (err) {
+              cb(err, address)
+            }
+          })
```

Comment:
I don't believe the branching is needed, the signature for the callback expected already matches.
```suggestion
          hookRunnerApplication('onListen', this[kAvvioBoot], this, function (err) {
              cb(err, address)
          })
```

Reply:
Hi, @metcoder95  we have tried implementing your changes and it resulted in many errors across many tests (either change does). We suspect that the callback for hookRunner only needs one that handles errors and otherwise causes trouble. Please let us know if we are wrong and maybe some other changes need to be made on top of the ones you have suggested.

Reply:
Doing a quick check over it, I'm guessing the error might be at [this line](https://github.com/fastify/fastify/pull/4899/files#diff-1b48bb429de959672dadab9c87c215e19177dd70e3bd932d131ce74294c44cd5R109). There the callback is getting called independently of what's the result of running `hookRunnerApplication`.

I also saw the same pattern for the `Promise` version, [here](https://github.com/fastify/fastify/pull/4899/files#diff-1b48bb429de959672dadab9c87c215e19177dd70e3bd932d131ce74294c44cd5R124-R127) the implementation is making the best effort to catch the `error`, but it is worthless because the promise was already resolved in either [L138](https://github.com/fastify/fastify/pull/4899/files#diff-1b48bb429de959672dadab9c87c215e19177dd70e3bd932d131ce74294c44cd5R138) or L134 (calling a `reject` when the `resolve` has been already called, results in `noop`).

Me, as a user, if using the `onListen` hook, I would expect that if anything fails within the hook, it is reported but does not stop the service.

It might be good to actually change that and just print a `this.logger.error`?

@Eomm @mcollina wdyt about? 🤔

Reply:
I disagree. 

If I write a hook handler, and i think that a failure is not critical, i just log it. If I think a failure is critical/fatal I throw it (pass it to done, whatever) and expect that the server exits.

I expect that ready hook handler behaves the same. 
Does an erroring ready hook handler just get logged?

Reply:
>If I write a application hook handler, and i think that a failure is not critical, i just log it. If I think a failure is critical/fatal I throw it (pass it to done, whatever) and expect that the server exits.

Yeah, that sounds right. Had a hard time deciding whether or not to stop the server, mostly because it is already listening. SGTM 👍 

If we follow that direction, we need to wait until `hookRunnerApplication` finishes iterating over the hooks. On the callback that handles the result of the `hookRunnerApplication`, if a failure happens, `this.close` should be called to stop the server from listening, and the promise should be rejected (or the callback should be called with the error as first parameter)

Reply:
I am trying to test this logic out, how should we wait for hookRunner to complete?

Reply:
We should wait for eomm and mcollina, because eomm meant that errors should be logged as warnings here https://github.com/fastify/fastify/pull/4580#discussion_r1103659263